treeherder/etl/taskcluster_pulse/handler.py

```diff
@@ -0,0 +1,302 @@
+    if not validateMessage(taskMetadata):
+        return None
+
+    # build a mapping from exchange name to status
+    eventMap = {
+      "exchange/taskcluster-queue/v1/task-pending": "pending",
+      "exchange/taskcluster-queue/v1/task-running": "running",
+      "exchange/taskcluster-queue/v1/task-completed": "completed",
+      "exchange/taskcluster-queue/v1/task-failed": "failed",
+      "exchange/taskcluster-queue/v1/task-exception": "exception",
+    }
+
+    taskType = eventMap[message["exchange"]]
+    job = None
+
+    if taskType == "pending":
+        runId = message["payload"]["runId"]
+        # If the task run was created for an infrastructure rerun, then resolve
+        # the previous run as retried.
+        if runId > 0:
+            handleTaskRerun(parsedRoute, taskMetadata, message["payload"])
+
+        job = handleTaskPending(parsedRoute, taskMetadata, message["payload"])
+    elif taskType == "running":
+        job = handleTaskRunning(parsedRoute, taskMetadata, message["payload"])
+    elif taskType == "completed":
+        job = handleTaskCompleted(parsedRoute, taskMetadata, message["payload"])
+    elif taskType == "failed":
+        job = handleTaskFailed(parsedRoute, taskMetadata, message["payload"])
+    elif taskType == "exception":
+        job = handleTaskException(parsedRoute, taskMetadata, message["payload"])
+    else:
+        raise Exception("Unknown exchange: {exchange}".format(exchange=message["exchange"]))
+
+    return job
+
+
+# Builds the basic Treeherder job message that's universal for all
+# messsage types.
+#
+# Specific handlers for each message type will add/remove information necessary
+# for the type of task event..
+def buildMessage(pushInfo, task, runId, message):
+    taskId = message["status"]["taskId"]
+    jobRun = message["status"]["runs"][runId]
+    treeherderConfig = task["extra"]["treeherder"]
+
+    job = {
+      "buildSystem": "taskcluster",
+      "owner": task["metadata"]["owner"],
+      "taskId": "{taskId}/{runId}".format(taskId=slugid.decode(taskId), runId=runId),
+      "retryId": runId,
+      "isRetried": False,
+      "display": {
+        # jobSymbols could be an integer (i.e. Chunk ID) but need to be strings
+        # for treeherder
+        "jobSymbol": str(treeherderConfig["symbol"]),
+        "groupSymbol": treeherderConfig.get("groupSymbol", "?"),
+        # Maximum job name length is 100 chars...
+        "jobName": task["metadata"]["name"][0:99],
+      },
+      "state": stateFromRun(jobRun),
+      "result": resultFromRun(jobRun),
+      "tier": treeherderConfig.get("tier", 1),
+      "timeScheduled": task["created"],
+      "jobKind": treeherderConfig.get("jobKind", "other"),
+      "reason": treeherderConfig.get("reason", "scheduled"),
+      "jobInfo": {
+        "links": [],
+        "summary": task["metadata"]["description"],
+      },
+      # XXX: The schema requires this value, however, I can't figure out from the code where this happens
```

Comment:
Good question!  I see the value in [Pulse Inspector](https://taskcluster-ui.herokuapp.com/pulse-messages?bindings%5B0%5D%5Bexchange%5D=exchange%2Ftaskcluster-treeherder%2Fv1%2Fjobs&bindings%5B0%5D%5Bpattern%5D=%23).  Poking around with `git grep` I finally found it at https://github.com/taskcluster/taskcluster/blob/7c73b2502e6f6a0836cea8dd27584face8d3494c/services/treeherder/src/exchanges.js#L48-L51

The version field could probably be dropped from the schema now that this message format is both produced and consumed in treeherder.  Or just set it here -- it doesn't hurt anything.

Reply:
I dropped it. Thanks for the find!